test/dotcom_web/live/subway_alerts_test.exs

```diff
@@ -0,0 +1,169 @@
+  alias Alerts.Alert
+  alias Dotcom.SystemStatus.Subway
+  alias Dotcom.Utils
+  alias DotcomWeb.Live.SubwayAlerts
+  alias Test.Support.Factories
+
+  @no_alerts_message "There are no other subway alerts at this time."
+
+  @service_impacting_effects service_impacting_effects() |> Enum.map(fn {effect, _} -> effect end)
+  @non_service_effects Alert.all_types() -- @service_impacting_effects
+
+  setup :verify_on_exit!
+
+  setup _ do
+    stub(Schedules.Repo.Mock, :end_of_rating, fn ->
+      Dotcom.Utils.ServiceDateTime.service_date() |> Timex.shift(days: 30)
+    end)
+
+    stub(Dotcom.Utils.DateTime.Mock, :coerce_ambiguous_date_time, fn date_time ->
+      Dotcom.Utils.DateTime.coerce_ambiguous_date_time(date_time)
+    end)
+
+    stub(Dotcom.Utils.DateTime.Mock, :now, fn ->
+      Dotcom.Utils.DateTime.now()
+    end)
+
+    stub(Alerts.Repo.Mock, :all, fn _ -> [] end)
+    stub(Alerts.Repo.Mock, :banner, fn -> nil end)
+    stub(Alerts.Repo.Mock, :by_route_ids, fn _, _ -> [] end)
+    stub(Routes.Repo.Mock, :by_type, fn _ -> [] end)
+
+    :ok
+  end
+
+  describe "subway alerts page" do
+    test "shows a message when there are no alerts", %{conn: conn} do
+      {:ok, view, _html} = live(conn, live_path(conn, SubwayAlerts))
+
+      assert render_async(view)
+             |> Floki.parse_document!()
+             |> Floki.get_by_id("station_and_service")
+             |> Floki.text() =~ @no_alerts_message
+    end
+
+    test "shows a route pill and no 'no alerts' message when an alert is present", %{conn: conn} do
+      route_id = Faker.Util.pick(Subway.lines())
+      route_type = Faker.Util.pick([0, 1])
+
+      stub(Alerts.Repo.Mock, :all, fn date ->
+        Factories.Alerts.Alert.build_list(1, :alert, %{
+          active_period: [Utils.ServiceDateTime.service_range_day(date)],
+          effect: Faker.Util.pick(@non_service_effects),
+          informed_entity:
+            Alerts.InformedEntitySet.new([
+              %Alerts.InformedEntity{
+                route_type: route_type,
+                route: route_id
+              }
+            ])
+        })
+      end)
+
+      stub(Routes.Repo.Mock, :by_type, fn [0, 1] ->
+        Factories.Routes.Route.build_list(1, :route, %{
+          id: route_id,
+          type: route_type
+        })
+      end)
+
+      {:ok, view, _html} = live(conn, live_path(conn, SubwayAlerts))
+
+      rendered_section =
+        render_async(view)
+        |> Floki.parse_document!()
+        |> Floki.get_by_id("station_and_service")
+
+      refute rendered_section |> Floki.text() =~ @no_alerts_message
+
+      refute rendered_section
+             |> Floki.find("[data-test=\"route_pill:#{route_id}\"]")
+             |> Enum.empty?()
+    end
+
+    test "does not put service-impacting alerts in the station_and_service section", %{conn: conn} do
+      route_id = Faker.Util.pick(Subway.lines())
+      route_type = Faker.Util.pick([0, 1])
+
+      stub(Alerts.Repo.Mock, :all, fn date ->
```

Comment:
Is there a reason you're stubbing these in the test instead of mocking them? With the `verify_on_exit!` above, I'm pretty sure it verifies mocks. Stubs don't get verified. The only reason I think we would want to use stubs in a test is if we don't know how many times the mock will be called.

Reply:
LiveView actually calls `mount` and `render` twice, once when the HTML is loaded and once again when the socket connection is established, so `expect` fails because the mocks are called twice. I could assert that the mocks are called twice, but I didn't for two reasons:
- It's a little confusing to see a random "this should be called twice", or "2 times however many times you actually expect it to be called" in the middle of an otherwise unrelated test.
- I wasn't 100% confident that the number of calls will always reliably be 2, and I wasn't sure it was worth the effort to confirm one way or the other.


I don't feel that strongly about it though - happy to switch it to `expect(yadda yadda, 2, :blah)` and then figure something else out if it turns out to be flaky or overly confusing.